toolchain/check/operator.cpp

```diff
@@ -0,0 +1,145 @@
+  // TODO: If the current package is the `Carbon` package, return
+  // `SemIR::InstId::Package`.
+
+  auto ident_id = context.identifiers().Lookup("Carbon");
+  if (!ident_id.is_valid()) {
+    return SemIR::NameScopeId::Invalid;
+  }
+  auto name_id = SemIR::NameId::ForIdentifier(ident_id);
+
+  // Look up `package.Carbon`.
+  auto package_id = context.LookupQualifiedName(
+      node_id, name_id, SemIR::NameScopeId::Package, /*required=*/false);
+  if (!package_id.is_valid()) {
+    return SemIR::NameScopeId::Invalid;
+  }
+
+  package_id = context.constant_values().Get(package_id).inst_id();
+
+  // We expect it to be a package, and fail if not.
+  if (auto package_inst =
+          context.insts().TryGetAs<SemIR::Namespace>(package_id)) {
+    auto& name_scope = context.name_scopes().Get(package_inst->name_scope_id);
+    // Check that this is really the `Carbon` package and not an alias.
+    if (name_scope.is_closed_import && name_scope.name_id == name_id &&
+        name_scope.enclosing_scope_id == SemIR::NameScopeId::Package) {
+      return package_inst->name_scope_id;
+    }
+  }
+  return SemIR::NameScopeId::Invalid;
+}
+
+// Returns the name scope of the operator interface for the specified operator
+// from the Carbon package.
+static auto GetOperatorInterface(Context& context, Parse::AnyExprId node_id,
+                                 Operator op) -> SemIR::NameScopeId {
+  auto carbon_package_id = GetCarbonPackage(context, node_id);
+  if (!carbon_package_id.is_valid()) {
+    return SemIR::NameScopeId::Invalid;
+  }
+
+  // Lookup `Carbon.InterfaceName`.
+  auto interface_ident_id = context.identifiers().Add(op.interface_name);
+  auto interface_id = context.LookupQualifiedName(
+      node_id, SemIR::NameId::ForIdentifier(interface_ident_id),
+      carbon_package_id, /*required=*/false);
+  if (!interface_id.is_valid()) {
+    return SemIR::NameScopeId::Invalid;
+  }
+
+  interface_id = context.constant_values().Get(interface_id).inst_id();
+
+  // We expect it to be an interface.
+  if (auto interface_inst =
+          context.insts().TryGetAs<SemIR::InterfaceType>(interface_id)) {
+    return context.interfaces().Get(interface_inst->interface_id).scope_id;
+  }
+  return SemIR::NameScopeId::Invalid;
+}
+
+// Returns the `Op` function for the specified operator.
+static auto GetOperatorOpFunction(Context& context, Parse::AnyExprId node_id,
+                                  Operator op) -> SemIR::InstId {
+  auto interface_scope_id = GetOperatorInterface(context, node_id, op);
+  if (!interface_scope_id.is_valid()) {
+    return SemIR::InstId::Invalid;
+  }
+
+  // Lookup `Interface.Op`.
+  auto op_ident_id = context.identifiers().Add(op.op_name);
+  auto op_id = context.LookupQualifiedName(
+      node_id, SemIR::NameId::ForIdentifier(op_ident_id), interface_scope_id,
+      /*required=*/false);
+  if (!op_id.is_valid()) {
+    return SemIR::InstId::Invalid;
+  }
+
+  op_id = context.constant_values().Get(op_id).inst_id();
```

Comment:
Similar to package_id comment, this feels odd.

Reply:
Comment added.